Hiring Data Analytics project

The hiring life cycle
- Number of candidates in each status over certain period.
- The average time on each status over certain period.

Failed candidates in each stage over certain period (CV Failed, technical interview failed OR management interview failed)

Rate of failure for each interviewer over certain period.

How many offers rejected by candidates over certain period (candidate reject the proposed offer).

Promising candidates for each account (CV passed, Ready to Call, Tech. interview passed and Ready to offer)


## logs json

In [2]:
import json

# Opening JSON file
f = open('logs.json')

# returns JSON object as
# a dictionary
data = json.load(f)

In [3]:
import json
import pandas as pd

# Load JSON data from file
with open('logs.json') as f:
    logs_data = json.load(f)

# Convert JSON data to a pandas DataFrame
df = pd.DataFrame(logs_data)

# Specify the desired fields for the Excel file
fields = ['ID', 'applicant_id', 'status', 'Created', 'log_user']

# Save the DataFrame to an Excel file
df.to_excel('logs_data.xlsx', index=False, columns=fields)


In [5]:
datalogs.head()

ID applicant_id                   status              Created log_user
0   5        2,496  0 - Record disqualified  01/30/2024 11:35 PM      NaN
1   6        2,463            0 - CV Failed  01/31/2024 12:17 PM      NaN
2   7        2,219            0 - CV Passed  01/31/2024 12:19 PM      NaN
3   8        2,462            0 - CV Passed  01/31/2024 12:24 PM      NaN
4   9        1,936            0 - CV Failed  01/31/2024 12:26 PM      NaN

In [4]:
datalogs=pd.read_excel("/content/logs_data.xlsx")

In [ ]:
data

[{'ID': 5,
  'applicant_id': '2,496',
  'status': '0 - Record disqualified',
  'Created': '01/30/2024 11:35 PM',
  'log_user': None},
 {'ID': 6,
  'applicant_id': '2,463',
  'status': '0 - CV Failed',
  'Created': '01/31/2024 12:17 PM',
  'log_user': None},
 {'ID': 7,
  'applicant_id': '2,219',
  'status': '0 - CV Passed',
  'Created': '01/31/2024 12:19 PM',
  'log_user': None},
 {'ID': 8,
  'applicant_id': '2,462',
  'status': '0 - CV Passed',
  'Created': '01/31/2024 12:24 PM',
  'log_user': None},
 {'ID': 9,
  'applicant_id': '1,936',
  'status': '0 - CV Failed',
  'Created': '01/31/2024 12:26 PM',
  'log_user': None},
 {'ID': 10,
  'applicant_id': '2,269',
  'status': '0 - CV Passed',
  'Created': '01/31/2024 12:31 PM',
  'log_user': None},
 {'ID': 11,
  'applicant_id': '2,439',
  'status': '0 - CV Passed',
  'Created': '01/31/2024 12:35 PM',
  'log_user': None},
 {'ID': 12,
  'applicant_id': '2,421',
  'status': '2 - Tech Interview Passed',
  'Created': '02/01/2024 10:03 AM',
  'l

## analyzing logs

In [1]:
import pandas as pd
logs_df=pd.read_excel("logs_data.xlsx")
logs_df.head()

ID applicant_id                   status              Created log_user
0   5        2,496  0 - Record disqualified  01/30/2024 11:35 PM      NaN
1   6        2,463            0 - CV Failed  01/31/2024 12:17 PM      NaN
2   7        2,219            0 - CV Passed  01/31/2024 12:19 PM      NaN
3   8        2,462            0 - CV Passed  01/31/2024 12:24 PM      NaN
4   9        1,936            0 - CV Failed  01/31/2024 12:26 PM      NaN

In [2]:
logs_df.shape

(943, 5)

In [3]:
logs_df.nunique()

ID              943
applicant_id    391
status           22
Created         645
log_user         21
dtype: int64

In [4]:
logs_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 943 entries, 0 to 942
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   ID            943 non-null    int64 
 1   applicant_id  943 non-null    object
 2   status        943 non-null    object
 3   Created       943 non-null    object
 4   log_user      920 non-null    object
dtypes: int64(1), object(4)
memory usage: 37.0+ KB


no duplicates since number of rows is 943 , number of unique id is 943

## Cleaning applicants

In [ ]:
import pandas as pd
applicants=pd.read_excel("Applicants.xlsx")
applicants.head()

/usr/local/lib/python3.10/dist-packages/google/colab/_dataframe_summarizer.py:88: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  cast_date_col = pd.to_datetime(column, errors="coerce")


ID      Applicant Name  Status ID  Role ID  Level ID Years of experience  \
0  2798        SAMAAN sadek          3      6.0       2.0                   2   
1  2797    Hegazy Aboelaref          1     11.0       1.0                   1   
2  2796      Hamdy Abu-Bakr         11      9.0       3.0                   7   
3  2795  SUHIL Wahbi Aghapy         11      9.0       3.0                   5   
4  2794         Rawan MAGDY          3     10.0       2.0                   5   

  Contact number                     emailAddress Technical interview date  \
0  0000-595-4035      samaan.sadek@somedomain.com                      NaN   
1  0110-020-1057  hegazy.aboelaref@somedomain.com                      NaN   
2  0122-547-1769    hamdy.abu-bakr@somedomain.com                      NaN   
3  0101-044-4294       suhil.wahbi@somedomain.com                      NaN   
4  0100-091-2923       rawan.magdy@somedomain.com                      NaN   

  Interviewers names Board interview date              Created By  \
0                NaN                  NaN      Abdelgaleel, Ahmad   
1                NaN                  NaN      Abdelgaleel, Ahmad   
2                NaN                  NaN      Abdelgaleel, Ahmad   
3                NaN                  NaN      Abdelgaleel, Ahmad   
4                NaN                  NaN  ibrahim, ahmed  farouk   

               Created             Modified By             Modified  \
0   4/17/2024 11:58 AM  Fawzy Bayomie, Mohamed   4/19/2024 10:07 AM   
1    4/15/2024 9:31 AM      Abdelgaleel, Ahmad    4/15/2024 9:31 AM   
2    4/14/2024 4:50 PM  Fawzy Bayomie, Mohamed   4/19/2024 10:09 AM   
3    4/14/2024 3:44 PM  Fawzy Bayomie, Mohamed   4/19/2024 10:09 AM   
4  2024-11-04 12:42:00  ibrahim, ahmed  farouk  2024-11-04 12:46:00   

     fit_to_account  
0                []  
1                []  
2  Alrajhi,Tawuniya  
3  Alrajhi,Tawuniya  
4                []

In [ ]:
applicants.tail()

ID                                   Applicant Name  Status ID  \
1895  177                        Elabedeen El-Hamid Shouab         18   
1896  172                             Sabri Iman EL-Hawary         13   
1897  170                         Zahran Akad Aly ELMANSEY         13   
1898  165                                   Shaltout Wadie         13   
1899  160  AL-KAFOURY IBRAHEEM MAHMOUD\nRAGAB\nHANAFY AZMY          2   

      Role ID  Level ID Years of experience Contact number  \
1895      NaN       NaN                  17  0110-579-8903   
1896      NaN       NaN                   0  0101-596-2745   
1897      NaN       NaN                   0  0180-020-8369   
1898      NaN       NaN                   8  0122-031-0901   
1899      NaN       NaN                   0  0102-942-4808   

                            emailAddress Technical interview date  \
1895   elabedeen.el-hamid@somedomain.com                      NaN   
1896           sabri.iman@somedomain.com       3/20/2018 11:00 PM   
1897          zahran.akad@somedomain.com       3/20/2018 11:00 PM   
1898       shaltout.wadie@somedomain.com      2018-10-07 21:30:00   
1899  al-kafoury.ibraheem@somedomain.com        3/19/2018 8:00 PM   

                   Interviewers names Board interview date  \
1895                              NaN                  NaN   
1896                              NaN                  NaN   
1897                              NaN                  NaN   
1898   Ezz El Deen, Ahmed;Fadl, Samar                  NaN   
1899  Fadl, Samar;Mansour, Abdel-Aziz                  NaN   

                  Created By              Created     Modified By  \
1895  Fawzy Bayomie, Mohamed  2018-05-06 14:32:00  Zafaan, Khaled   
1896  Fawzy Bayomie, Mohamed  2018-05-06 14:32:00  Zafaan, Khaled   
1897  Fawzy Bayomie, Mohamed  2018-05-06 14:32:00  Zafaan, Khaled   
1898  Fawzy Bayomie, Mohamed  2018-05-06 14:32:00  Zafaan, Khaled   
1899  Fawzy Bayomie, Mohamed  2018-05-06 14:32:00   Hussein, Mona   

                 Modified fit_to_account  
1895  2019-04-03 17:08:00             []  
1896    6/27/2018 4:42 PM             []  
1897    6/27/2018 4:43 PM             []  
1898  2018-12-07 16:48:00             []  
1899   4/18/2019 12:09 PM             []

In [ ]:
applicants.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1900 entries, 0 to 1899
Data columns (total 16 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   ID                        1900 non-null   int64  
 1   Applicant Name            1899 non-null   object 
 2   Status ID                 1900 non-null   int64  
 3   Role ID                   1475 non-null   float64
 4   Level ID                  1639 non-null   float64
 5   Years of experience       1858 non-null   object 
 6   Contact number            1762 non-null   object 
 7   emailAddress              1900 non-null   object 
 8   Technical interview date  658 non-null    object 
 9   Interviewers names        557 non-null    object 
 10  Board interview date      144 non-null    object 
 11  Created By                1900 non-null   object 
 12  Created                   1900 non-null   object 
 13  Modified By               1900 non-null   object 
 14  Modified

In [ ]:
# Detecting duplicates
duplicates = applicants[applicants.duplicated()]
duplicates


Empty DataFrame
Columns: [ID, Applicant Name, Status ID, Role ID, Level ID, Years of experience, Contact number, emailAddress, Technical interview date, Interviewers names, Board interview date, Created By, Created, Modified By, Modified, fit_to_account]
Index: []

In [ ]:
applicants.nunique()

ID                          1900
Applicant Name              1899
Status ID                     26
Role ID                       20
Level ID                       5
Years of experience           22
Contact number              1762
emailAddress                1888
Technical interview date     603
Interviewers names           164
Board interview date         131
Created By                    38
Created                     1595
Modified By                   43
Modified                    1324
fit_to_account                61
dtype: int64

In [ ]:
# change data type of dates

applicants['Technical interview date'] = pd.to_datetime(applicants['Technical interview date'])
applicants['Board interview date'] = pd.to_datetime(applicants['Board interview date'])
applicants['Created'] = pd.to_datetime(applicants['Created'])
applicants['Modified'] = pd.to_datetime(applicants['Modified'])

<ipython-input-3-8a50a8d3402c>:4: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  applicants['Board interview date'] = pd.to_datetime(applicants['Board interview date'])


In [ ]:
applicants.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1900 entries, 0 to 1899
Data columns (total 16 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   ID                        1900 non-null   int64         
 1   Applicant Name            1899 non-null   object        
 2   Status ID                 1900 non-null   int64         
 3   Role ID                   1475 non-null   float64       
 4   Level ID                  1639 non-null   float64       
 5   Years of experience       1858 non-null   object        
 6   Contact number            1762 non-null   object        
 7   emailAddress              1900 non-null   object        
 8   Technical interview date  658 non-null    datetime64[ns]
 9   Interviewers names        557 non-null    object        
 10  Board interview date      144 non-null    datetime64[ns]
 11  Created By                1900 non-null   object        
 12  Created             

In [ ]:
# Check for duplicate rows
duplicate_rows = applicants[applicants.duplicated()]

# Display duplicate rows (if any)
if not duplicate_rows.empty:
    print("Duplicate Rows:")
    print(duplicate_rows)
else:
    print("No duplicate rows found.")

No duplicate rows found.


In [ ]:
applicants.isnull().sum()

ID                             0
Applicant Name                 1
Status ID                      0
Role ID                      425
Level ID                     261
Years of experience           42
Contact number               138
emailAddress                   0
Technical interview date    1242
Interviewers names          1343
Board interview date        1756
Created By                     0
Created                        0
Modified By                    0
Modified                       0
fit_to_account                 0
dtype: int64

In [ ]:
applicants['Technical interview date'] = applicants['Technical interview date'].fillna('1900-01-01')
applicants['Board interview date'] = applicants['Board interview date'].fillna('1900-01-01')
applicants['Created'] = applicants['Created'].fillna('1900-01-01')
applicants['Modified'] = applicants['Modified'].fillna('1900-01-01')

In [ ]:
applicants.to_excel('cleaned_applicants_final.xlsx', index=False)

In [ ]:
applicants.head()

ID      Applicant Name  Status ID  Role ID  Level ID Years of experience  \
0  2798        SAMAAN sadek          3      6.0       2.0                   2   
1  2797    Hegazy Aboelaref          1     11.0       1.0                   1   
2  2796      Hamdy Abu-Bakr         11      9.0       3.0                   7   
3  2795  SUHIL Wahbi Aghapy         11      9.0       3.0                   5   
4  2794         Rawan MAGDY          3     10.0       2.0                   5   

  Contact number                     emailAddress Technical interview date  \
0  0000-595-4035      samaan.sadek@somedomain.com               1900-01-01   
1  0110-020-1057  hegazy.aboelaref@somedomain.com               1900-01-01   
2  0122-547-1769    hamdy.abu-bakr@somedomain.com               1900-01-01   
3  0101-044-4294       suhil.wahbi@somedomain.com               1900-01-01   
4  0100-091-2923       rawan.magdy@somedomain.com               1900-01-01   

  Interviewers names Board interview date              Created By  \
0                NaN           1900-01-01      Abdelgaleel, Ahmad   
1                NaN           1900-01-01      Abdelgaleel, Ahmad   
2                NaN           1900-01-01      Abdelgaleel, Ahmad   
3                NaN           1900-01-01      Abdelgaleel, Ahmad   
4                NaN           1900-01-01  ibrahim, ahmed  farouk   

              Created             Modified By            Modified  \
0 2024-04-19 10:07:00  Fawzy Bayomie, Mohamed 2024-04-19 10:07:00   
1 2024-04-15 09:31:00      Abdelgaleel, Ahmad 2024-04-15 09:31:00   
2 2024-04-19 10:09:00  Fawzy Bayomie, Mohamed 2024-04-19 10:09:00   
3 2024-04-19 10:09:00  Fawzy Bayomie, Mohamed 2024-04-19 10:09:00   
4 2024-11-04 12:46:00  ibrahim, ahmed  farouk 2024-11-04 12:46:00   

     fit_to_account  
0                []  
1                []  
2  Alrajhi,Tawuniya  
3  Alrajhi,Tawuniya  
4                []

In [ ]:
applicants.to_excel('cleaned_applicants.xlsx', index=False)


In [ ]:
applicants_mod=pd.read_excel("cleaned_applicants.xlsx")
applicants_mod.head()

ID      Applicant Name  Status ID  Role ID  Level ID Years of experience  \
0  2798        SAMAAN sadek          3      6.0       2.0                   2   
1  2797    Hegazy Aboelaref          1     11.0       1.0                   1   
2  2796      Hamdy Abu-Bakr         11      9.0       3.0                   7   
3  2795  SUHIL Wahbi Aghapy         11      9.0       3.0                   5   
4  2794         Rawan MAGDY          3     10.0       2.0                   5   

  Contact number                     emailAddress Technical interview date  \
0  0000-595-4035      samaan.sadek@somedomain.com                      NaN   
1  0110-020-1057  hegazy.aboelaref@somedomain.com                      NaN   
2  0122-547-1769    hamdy.abu-bakr@somedomain.com                      NaN   
3  0101-044-4294       suhil.wahbi@somedomain.com                      NaN   
4  0100-091-2923       rawan.magdy@somedomain.com                      NaN   

  Interviewers names              Created By              Created  \
0            Missing      Abdelgaleel, Ahmad   4/17/2024 11:58 AM   
1            Missing      Abdelgaleel, Ahmad    4/15/2024 9:31 AM   
2            Missing      Abdelgaleel, Ahmad    4/14/2024 4:50 PM   
3            Missing      Abdelgaleel, Ahmad    4/14/2024 3:44 PM   
4            Missing  ibrahim, ahmed  farouk  2024-11-04 12:42:00   

              Modified By             Modified    fit_to_account  
0  Fawzy Bayomie, Mohamed   4/19/2024 10:07 AM                []  
1      Abdelgaleel, Ahmad    4/15/2024 9:31 AM                []  
2  Fawzy Bayomie, Mohamed   4/19/2024 10:09 AM  Alrajhi,Tawuniya  
3  Fawzy Bayomie, Mohamed   4/19/2024 10:09 AM  Alrajhi,Tawuniya  
4  ibrahim, ahmed  farouk  2024-11-04 12:46:00                []

In [ ]:
# Count the number of candidates in each status
status_counts = applicants['Status ID'].value_counts()  # Assuming 'Status ID' is the column containing status information

# Display the count for each status
print("Number of Candidates in Each Status:")
print(status_counts)


Number of Candidates in Each Status:
Status ID
3     326
1     317
13    260
26    193
8     136
4     125
24     98
17     86
18     62
23     61
25     42
2      36
12     25
20     25
7      19
14     18
16     16
10     15
11     12
15      6
5       6
6       5
22      4
21      4
19      2
9       1
Name: count, dtype: int64


In [ ]:
modified_applicants['fit_to_account'].nunique()

61

In [ ]:
modified_applicants['fit_to_account'].unique()

array(['None', 'Alrajhi,Tawuniya', 'MLI', 'MCIT EG', 'MES,MCIT EG',
       'Tawuniya', 'Tawuniya,MCIT EG', 'Alrajhi,MCIT EG',
       'MCIT EG,Tawuniya', 'MES,VFDE Infra', 'VFDE Infra', 'Alrajhi',
       'Alrajhi,Early Career', 'MCIT EG,MES', 'MES', 'Man Utd,MES',
       'Early Career,Alrajhi', 'Tawuniya,Alrajhi,MCIT EG', 'Alrajhi,LM',
       'LM', 'VFDE Infra,MES', 'VFDE Infra,MCIT EG',
       'LM,Tawuniya,Alrajhi', 'BMW', 'VFDE Infra,MES,Alrajhi',
       'LM,Alrajhi,MCIT EG', 'Early Career', 'LM,VFDE Infra,MCIT EG',
       'LM,MCIT EG', 'VFDE Infra,MES,Volkswagen AG,MCIT EG', 'AAIB',
       'VFDE Infra,DCMP,Alrajhi', 'LM,Alrajhi', 'MCIT EG,Alrajhi',
       'Man Utd', 'LM,Man Utd', 'LM,MES', 'Man Utd,LM',
       'VFDE Infra,Man Utd', 'Alrajhi,VFDE Infra', 'LM,InGrid',
       'VFDE Infra,LM', 'Saudi Payment', 'LM,DCMP,InGrid', 'InGrid,BMW',
       'LM,VFDE Infra', 'LM,VFDE Infra,InGrid,MES', 'MES,LM',
       'VFDE Infra,LM,DCMP', 'VFDE Infra,AAIB', 'LM,AAIB', 'P&G',
       'InGrid,AAIB'

In [ ]:
modified_applicants['Years of experience'].nunique()

array(['2', '1', '7', '5', '0', '14', '10', '3', '9', '6', '11', '4',
       '15', '17', '8', '19', '16', '13', '20+', '12', '20', '18', nan],
      dtype=object)

## Avalibale roles

In [ ]:
available_roles=pd.read_csv("Available Roles.csv")
available_roles.head()

Roles  ID
0    AMS Linux   1
1  AMS Windows   2
2  AI Engineer   3
3          AWS   4
4        Azure   5

In [ ]:
# Check for duplicate rows
duplicate_rows = available_roles[available_roles.duplicated()]

# Display duplicate rows (if any)
if not duplicate_rows.empty:
    print("Duplicate Rows:")
    print(duplicate_rows)
else:
    print("No duplicate rows found.")

No duplicate rows found.


In [ ]:
available_roles.nunique()

Roles    21
ID       21
dtype: int64

In [ ]:
available_roles.isnull().sum()

Roles    0
ID       0
dtype: int64

In [ ]:
available_roles.shape

(1900, 16)

## Expected level

In [ ]:
import pandas as pd

expected_level=pd.read_csv("Expected Level.csv")
expected_level.head()

Expected level  ID
0          Entry   1
1   Intermediate   2
2     Specialist   3
3         Expert   4
4         Master   5

In [ ]:
expected_level.nunique()

Expected level    5
ID                5
dtype: int64

In [ ]:
expected_level.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Expected level  5 non-null      object
 1   ID              5 non-null      int64 
dtypes: int64(1), object(1)
memory usage: 208.0+ bytes


In [ ]:
expected_level.isnull().sum()

Expected level    0
ID                0
dtype: int64

In [ ]:
expected_level.shape

(5, 2)

## Status

In [ ]:
status=pd.read_csv("Status.csv")
status

Status  ID
0                    0 - CV Added   1
1         0 - Record disqualified   2
2                   0 - CV Passed   3
3                   0 - CV Failed   4
4               0 - Not Reachable   5
5            0- Call Short-listed   6
6                 1 - Call Passed   7
7                 1 - Call Failed   8
8                  1- Exam Passed   9
9                  1- Exam Failed  10
10              1 - Ready To Call  11
11      2 - Tech Interview Passed  12
12      2 - Tech Interview Failed  13
13                    2 - On hold  14
14   3 - Board Interview Deferred  15
15     3 - Board Interview Passed  16
16     3 - Board Interview Failed  17
17                    3 - On hold  18
18  4-Ready For Account Interview  19
19             4 - Ready to Offer  20
20                    4 - Offered  21
21             4 - Offer Accepted  22
22             4 - Offer Rejected  23
23                      5 - Hired  24
24                5 - Blacklisted  25
25             5 - Not interested  26

In [ ]:
status.nunique()

Status    26
ID        26
dtype: int64

In [ ]:
status.describe()

ID
count  26.000000
mean   13.500000
std     7.648529
min     1.000000
25%     7.250000
50%    13.500000
75%    19.750000
max    26.000000

In [ ]:
status.isnull().sum()

Status    0
ID        0
dtype: int64

In [ ]:
status.shape

(26, 2)